Wrappers/Python/cil/optimisation/utilities/preconditioner.py

```diff
@@ -0,0 +1,199 @@
+        """
+        self.array = self.operator.adjoint(
+            self.operator.range_geometry().allocate(value=1.0))
+        
+        self.operator.range_geometry().allocate(value=1.0).divide(
+                self.array, where=np.abs(self.array.as_array()) > 0, out=self.array)
+
+    def apply(self, algorithm, gradient, out=None):
+        r"""
+        Update the preconditioner.
+
+        Parameters
+        ----------
+        algorithm : object
+            The algorithm object.
+        gradient : DataContainer
+            The calculated gradient to modify
+        out : DataContainer, 
+            Container to fill with the modified gradient 
+            
+        Returns
+        -------
+        DataContainer
+            The modified gradient
+            
+        """
+        if out is None:
+            out = gradient.copy()
+        gradient.multiply(
+            self.array, out=out)
+        return out
+
+
+class AdaptiveSensitivity(Sensitivity):
+
+    r"""
+    Adaptive Sensitivity preconditioner class. 
+
+    In each call to the preconditioner the `gradient` is multiplied by :math:` (x+\delta) /(A^T \mathbf{1})` where :math:`A` is an operator,  :math:`\mathbf{1}` is an object in the range of the operator filled with ones.
+    The point :math:`x` is the current iteration, or a reference image,  and :math:`delta` is a small positive float. 
+
+
+    Parameters
+    ----------
+    operator : CIL object
+        The operator used for sensitivity computation.
+    delta : float, optional
+        The delta value for the preconditioner.
+    reference : DataContainer e.g. ImageData, default is None
+        Reference data, an object in the domain of the operator. Recommended to be a best guess reconstruction. If reference data is passed the preconditioner does not update with each iteration.  
+    iterations : int,  default = 100
+        The maximum number of iterations before the preconditoner is frozen and no-longer updates. Note that if reference data is passed the preconditioner is always frozen and `iterations` is set to -1. 
+    
+    Reference
+    ---------
+    Twyman R, Arridge S, Kereta Z, Jin B, Brusaferri L, Ahn S, Stearns CW, Hutton BF, Burger IA, Kotasidis F, Thielemans K. An Investigation of Stochastic Variance Reduction Algorithms for Relative Difference Penalized 3D PET Image Reconstruction. IEEE Trans Med Imaging. 2023 Jan;42(1):29-41. doi: 10.1109/TMI.2022.3203237. Epub 2022 Dec 29. PMID: 36044488.
+
+    """
+
+    def __init__(self, operator, delta=1e-6, iterations=100, reference=None):
+
+        self.iterations = iterations
+        self.delta = delta
+        
+        super(AdaptiveSensitivity, self).__init__(
+            operator=operator)
+        
+        
+        self.freezing_point = operator.domain_geometry().allocate(0)
+        if reference is not None:
+              self.array.multiply(reference+self.delta, out=self.freezing_point)
```

Comment:
This might save some memory but cost in computation

```suggestion
              reference += self.delta
              self.array.multiply(reference, out=self.freezing_point)
              reference -= self.delta
```